// @apollo/space-kit/FeatureIntroLearnMoreLink/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
require('../fonts/fonts.js');
var base$1 = require('../base-91f9e1df.js');
require('../mono-a0d4bd30.js');
var colors = require('../colors/colors.js');
require('../assertUnreachable-698a501a.js');
var createElementFromAs = require('../createElementFromAs-30a442da.js');
var index = require('../index-8a508515.js');

/**
 * Component to render Feature Intro learn more button.
 *
 * If this component is rendered in the children of `FeatureIntroControl`, then
 * `FeatureIntroControl` will render this element in it's layout. Otherwise, it's
 * rendered as-is.
 */

var FeatureIntroLearnMoreLink = function (_a) {
  var _b = _a.as,
      as = _b === void 0 ? "a" : _b,
      className = _a.className,
      style = _a.style;
  var featureIntroContext = index.useFeatureIntroControlInternalContext();
  var _c = [featureIntroContext === null || featureIntroContext === void 0 ? void 0 : featureIntroContext.id, featureIntroContext === null || featureIntroContext === void 0 ? void 0 : featureIntroContext.setLearnMoreLink],
      featureIntroId = _c[0],
      setLearnMoreLink = _c[1];
  var element = React.useMemo(function () {
    return /*#__PURE__*/React.isValidElement(as) ? /*#__PURE__*/React.cloneElement(createElementFromAs.createElementFromAs(as), {
      className: className,
      style: style
    }, core.jsx("div", {
      id: featureIntroId && featureIntroId + "-learn-more-link"
    }, core.jsx("div", {
      css: _tslib.__assign(_tslib.__assign({
        color: colors.colors.grey.base
      }, base$1.small), {
        fontWeight: 600,
        padding: "8px 5px",
        marginLeft: -5,
        marginTop: -8
      })
    }, "Learn more"))) : null;
  }, [as, className, featureIntroId, style]);
  React.useLayoutEffect(function () {
    setLearnMoreLink === null || setLearnMoreLink === void 0 ? void 0 : setLearnMoreLink(element);
  }, [element, setLearnMoreLink]);
  return featureIntroContext ? null : element;
};

exports.FeatureIntroLearnMoreLink = FeatureIntroLearnMoreLink;
