// @apollo/space-kit/FormHelperText/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
require('../fonts/fonts.js');
var base$1 = require('../base-91f9e1df.js');
require('../mono-a0d4bd30.js');
var colors = require('../colors/colors.js');
var index = require('../index-14c98598.js');
var IconInfo = require('../icons/IconInfo.js');

/**
 * Component to render a helper text
 *
 * This is intended to be rendered below `<FormControl>`. If this is rendered on
 * it's own; it will render `children` without any modification.
 */

var FormHelperText = function (_a) {
  var children = _a.children,
      className = _a.className,
      _b = _a.showIcon,
      showIcon = _b === void 0 ? false : _b;
  var setHelper = index.useFormControlInternalContext().setHelper;
  var element = React.useMemo(function () {
    return core.jsx("div", {
      className: className,
      css: core.css(_tslib.__assign(_tslib.__assign({}, base$1.small), {
        color: colors.colors.grey.base,
        display: "flex"
      }))
    }, showIcon ? core.jsx(IconInfo.IconInfo, {
      css: {
        height: 15,
        marginRight: 8,
        position: "relative",
        top: 2,
        width: 15
      }
    }) : null, core.jsx("div", null, children));
  }, [className, showIcon, children]);
  React.useLayoutEffect(function () {
    // This will cause a bug if you change the `error` prop
    setHelper === null || setHelper === void 0 ? void 0 : setHelper(element);
    return function () {
      return setHelper === null || setHelper === void 0 ? void 0 : setHelper(null);
    };
  }, [element, setHelper]); // If `setHelper` exists then we're rendering this under the form control
  // context provider. `FormControl` will pull that element from the context and
  // insert into the layout, so return `null`.

  return setHelper ? null : element;
};

exports.FormHelperText = FormHelperText;
