// @apollo/space-kit/Tooltip/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
require('../fonts/fonts.js');
require('../base-91f9e1df.js');
require('../mono-a0d4bd30.js');
require('../colors/colors.js');
require('../SpaceKitProvider/index.js');
require('classnames');
var TooltipContext = require('../TooltipContext-20cc2d3d.js');
require('../SingletonComponent-7ca6baa8.js');
var index$1 = require('../index-6241ea02.js');
require('@tippyjs/react');
var mouseRestPlugin = require('./tooltip/mouseRestPlugin.js');

/**
 * Tooltip
 *
 * Intended to show non-essential, non-interactive information after a short
 * delay when hovering on an element. We also wait for the user's cursor to rest
 * on the element; meaning we won't show a Tooltip while the cursor is moving
 */

var Tooltip = function (_a) {
  var children = _a.children,
      visible = _a.visible,
      props = _tslib.__rest(_a, ["children", "visible"]);

  return /*#__PURE__*/React__default.createElement(TooltipContext.TooltipContextProvider, {
    descendsFromTooltip: true
  }, /*#__PURE__*/React__default.createElement(index$1.AbstractTooltip, _tslib.__assign({
    delay: 150,
    plugins: [mouseRestPlugin.mouseRest]
  }, typeof visible === "boolean" ? {
    visible: visible
  } : {
    trigger: "mouseenter"
  }, props), children));
};

exports.Tooltip = Tooltip;
