// @apollo/space-kit/FeatureIntroImage/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
var index = require('../index-8a508515.js');

/** @jsx jsx */
/**
 * Component to render Feature Intro Image.
 *
 * If this component is rendered in the children of `FeatureIntroControl`, then
 * `FeatureIntroControl` will render this element in it's layout. Otherwise, it's
 * rendered as-is.
 */

var FeatureIntroImage = function (_a) {
  var children = _a.children,
      className = _a.className,
      style = _a.style;
  var featureIntroContext = index.useFeatureIntroControlInternalContext();
  var _b = [featureIntroContext === null || featureIntroContext === void 0 ? void 0 : featureIntroContext.id, featureIntroContext === null || featureIntroContext === void 0 ? void 0 : featureIntroContext.setImage],
      featureIntroId = _b[0],
      setImage = _b[1];
  var element = React.useMemo(function () {
    return core.jsx("div", {
      className: className,
      css: {
        flexBasis: "50%"
      },
      id: featureIntroId + "-image",
      style: style
    }, children);
  }, [children, className, featureIntroId, style]);
  React.useLayoutEffect(function () {
    setImage === null || setImage === void 0 ? void 0 : setImage(element);
  }, [element, setImage]);
  return featureIntroContext ? null : element;
};

exports.FeatureIntroImage = FeatureIntroImage;
