// @apollo/space-kit/FormStartAdornment/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
var index = require('../index-14c98598.js');

/** @jsx jsx */
/**
 * Component to render a start adornment.
 *
 * This is intended to be rendered below `<FormControl>`. If this is rendered on
 * it's own; it will render `children` without any modification.
 */

var FormStartAdornment = function (_a) {
  var children = _a.children,
      className = _a.className;
  var setStartAdornment = index.useFormControlInternalContext().setStartAdornment;
  var element = React.useMemo(function () {
    return core.jsx("div", {
      css: core.css({
        position: "absolute",
        display: "inline-flex",
        left: 12,
        top: "50%",
        transform: "translateY(-50%)"
      }),
      className: className
    }, children);
  }, [children, className]);
  React.useLayoutEffect(function () {
    setStartAdornment === null || setStartAdornment === void 0 ? void 0 : setStartAdornment(element);
    return function () {
      return setStartAdornment === null || setStartAdornment === void 0 ? void 0 : setStartAdornment(null);
    };
  }, [element, setStartAdornment]);
  return setStartAdornment ? null : element;
};

exports.FormStartAdornment = FormStartAdornment;
