// @apollo/space-kit/FeatureIntroDismissButton/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
require('../fonts/fonts.js');
var base$1 = require('../base-91f9e1df.js');
require('../mono-a0d4bd30.js');
var colors = require('../colors/colors.js');
var index = require('../index-8a508515.js');

/**
 * Component to render Feature Intro content.
 *
 * If this component is rendered in the children of `FeatureIntroControl`, then
 * `FeatureIntroControl` will render this element in it's layout. Otherwise, it's
 * rendered as-is.
 */

var FeatureIntroDismissButton = function (_a) {
  var children = _a.children,
      className = _a.className,
      style = _a.style,
      onClick = _a.onClick;
  var featureIntroContext = index.useFeatureIntroControlInternalContext();
  var _b = [featureIntroContext === null || featureIntroContext === void 0 ? void 0 : featureIntroContext.id, featureIntroContext === null || featureIntroContext === void 0 ? void 0 : featureIntroContext.setDismissButton],
      featureIntroId = _b[0],
      setDismissButton = _b[1];
  var element = React.useMemo(function () {
    return core.jsx("div", {
      className: className,
      css: _tslib.__assign(_tslib.__assign({}, base$1.small), {
        color: colors.colors.blue.base,
        fontWeight: 600,
        padding: "8px 5px",
        marginRight: -5,
        marginTop: -8,
        // force right even if there is no learn more button
        marginLeft: "auto",
        cursor: "pointer"
      }),
      id: featureIntroId && featureIntroId + "-dismiss-button",
      style: style,
      onClick: onClick
    }, children);
  }, [children, className, featureIntroId, onClick, style]);
  React.useLayoutEffect(function () {
    setDismissButton === null || setDismissButton === void 0 ? void 0 : setDismissButton(element);
  }, [element, setDismissButton]);
  return featureIntroContext ? null : element;
};

exports.FeatureIntroDismissButton = FeatureIntroDismissButton;
