// @apollo/space-kit/FormEndAdornment/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
var index = require('../index-14c98598.js');

/** @jsx jsx */
/**
 * Component to render a end adornment.
 *
 * This is intended to be rendered below `<FormControl>`. If this is rendered on
 * it's own; it will render `children` without any modification.
 */

var FormEndAdornment = function (_a) {
  var children = _a.children,
      className = _a.className;
  var setEndAdornment = index.useFormControlInternalContext().setEndAdornment;
  var element = React.useMemo(function () {
    return core.jsx("div", {
      css: core.css({
        display: "inline-flex",
        position: "absolute",
        right: 12,
        textAlign: "right",
        top: "50%",
        transform: "translateY(-50%)"
      }),
      className: className
    }, children);
  }, [children, className]);
  React.useLayoutEffect(function () {
    setEndAdornment === null || setEndAdornment === void 0 ? void 0 : setEndAdornment(element);
    return function () {
      return setEndAdornment === null || setEndAdornment === void 0 ? void 0 : setEndAdornment(null);
    };
  }, [element, setEndAdornment]);
  return setEndAdornment ? null : element;
};

exports.FormEndAdornment = FormEndAdornment;
