// @apollo/space-kit/ListDivider/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
var colors = require('../colors/colors.js');
var assertUnreachable = require('../assertUnreachable-698a501a.js');
require('tinycolor2');
var index = require('../index-497a4b3b.js');
require('../verticalListMarginFromPadding-9f8e64e0.js');
var index$1 = require('../List/index.js');
var index$2 = require('../ListItem/index.js');

/**
 * Divider between sections in a list
 */

var ListDivider = function (props) {
  var listConfig = index.useListConfig(); // Stop click events so we don't try to close the list when clicking something
  // non-interactive

  var handleClick = React__default.useCallback(function (event) {
    event.preventDefault();
    event.stopPropagation();
  }, []);
  /**
   * Indicates how much negative space we need to take to make sure the divider
   * goes edge-to-edge
   */

  var horizontalOffset = listConfig.margin === "auto" ? -index$1.listPadding : listConfig.margin === "none" ? 0 : assertUnreachable.assertUnreachable(listConfig.margin);
  return core.jsx(index.ListConfigProvider, {
    truncate: false
  }, core.jsx(index$2.ListItem, {
    interactive: false,
    css: core.css({
      alignItems: "stretch",
      justifyItems: "center",
      paddingLeft: 0,
      paddingRight: 0,
      position: "relative",
      height: 17
    })
  }, core.jsx("hr", _tslib.__assign({}, props, {
    onClick: handleClick,
    css: {
      borderBottomWidth: 0,
      borderLeftWidth: 0,
      borderRightWidth: 0,
      borderTopColor: colors.colors.silver.base,
      borderTopStyle: "solid",
      borderTopWidth: 1,
      bottom: 0,
      left: horizontalOffset,
      margin: 0,
      position: "absolute",
      right: horizontalOffset,
      top: 8
    }
  }))));
};

exports.ListDivider = ListDivider;
