// @apollo/space-kit/index-8a508515.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

/**
 * Context holding all configuration options for feature intro
 */

var FeatureIntroControlContext = /*#__PURE__*/React.createContext(undefined);
var FeatureIntroControlContextProvider = function (_a) {
  var children = _a.children,
      id = _a.id;

  var _b = React.useState(),
      content = _b[0],
      setContent = _b[1];

  var _c = React.useState(),
      heading = _c[0],
      setHeading = _c[1];

  var _d = React.useState(),
      image = _d[0],
      setImage = _d[1];

  var _e = React.useState(),
      learnMoreLink = _e[0],
      setLearnMoreLink = _e[1];

  var _f = React.useState(),
      dismissButton = _f[0],
      setDismissButton = _f[1];

  var contextValue = React.useMemo(function () {
    return {
      id: id,
      content: content,
      setContent: setContent,
      heading: heading,
      setHeading: setHeading,
      image: image,
      setImage: setImage,
      learnMoreLink: learnMoreLink,
      setLearnMoreLink: setLearnMoreLink,
      dismissButton: dismissButton,
      setDismissButton: setDismissButton
    };
  }, [content, dismissButton, heading, id, image, learnMoreLink]);
  return /*#__PURE__*/React.createElement(FeatureIntroControlContext.Provider, {
    value: contextValue
  }, children);
};
/**
 * Internal hook to access Feature Intro control context
 */

function useFeatureIntroControlInternalContext() {
  return React.useContext(FeatureIntroControlContext);
}

exports.FeatureIntroControlContextProvider = FeatureIntroControlContextProvider;
exports.useFeatureIntroControlInternalContext = useFeatureIntroControlInternalContext;
